src/ra_server.erl

```diff
@@ -3314,6 +3443,186 @@ after_log_append_reply(Cmd, Idx, Term, Effects0) ->
+            %% This node is alone in the cluster, we can send the `noop'
+            %% command with the newer machine version right away.
+            MacVer = ra_machine:version(Mac),
+            Noop = {noop, #{ts => erlang:system_time(millisecond)},
+                    MacVer},
+            NoopEffect = {next_event, cast, {command, Noop}},
+            [NoopEffect];
+        _ ->
+            %% We continue to send the `noop' command immediately after
+            %% becoming a leader, but compared to Ra 2.15 and older, we don't
+            %% set the machine version to the latest: we keep the same
+            %% effective machine version for now.
+            %%
+            %% However, we query info keys from all peers, including their
+            %% supported machine version. The replies will be used to
+            %% determine the max supported machine version and when it is
+            %% greater than the effective one, another `noop' command will be
+            %% sent.
+            Noop = {noop, #{ts => erlang:system_time(millisecond)},
+                    EffectiveMacVer},
+            NoopEffect = {next_event, cast, {command, Noop}},
+            InfoRpcEffects = info_rpc_effects(State),
+            [NoopEffect | InfoRpcEffects]
+    end.
+
+info_rpc_effects(#{cfg := #cfg{id = Id}, cluster := Cluster} = State) ->
+    InfoRpcEffects = maps:fold(
+                       fun
+                           (PeerId, _, Acc) when PeerId =:= Id ->
+                               Acc;
+                           (PeerId, _, Acc) ->
+                               Acc ++ info_rpc_effects_for_peer(State, PeerId)
+                       end, [], Cluster),
+    InfoRpcEffects.
+
+info_rpc_effects_for_peer(
+  #{cluster := Cluster, current_term := CurTerm} = State, PeerId) ->
+    %% We determine if we need to ask (for the fist time or again) the info
+    %% from a peer.
+    SendRpc = case Cluster of
+                  #{PeerId := #{machine_version := PeerMacVer}} ->
+                      %% We have the machine version of the peer, but we want
+                      %% to ask again if that version is old, in the hope the
+                      %% peer restarted and was updated.
+                      MacVer = machine_version(State),
+                      PeerMacVer < MacVer;
+                  _ ->
+                      %% We don't have any details about the peer, we ask.
+                      true
+              end,
+    case SendRpc of
+        true ->
+            %% We ask for all info keys currently. If we ask for specific
+            %% keys, we will have to handle merging of already known info keys
+            %% and updates.
+            Id = id(State),
+            Command = #info_rpc{from = Id,
+                                term = CurTerm,
+                                keys = [machine_version]},
+            [{send_rpc, PeerId, Command}];
+        false ->
+            []
+    end.
+
+info_reply_effect(
```

Comment:
this should fit to the right of `(` so no need for newline here.

Reply:
Done